bot/exts/info/doc/_redis_cache.py

```diff
@@ -0,0 +1,87 @@
+from __future__ import annotations
+
+import datetime
+import pickle
+from typing import Optional, TYPE_CHECKING
+
+from async_rediscache.types.base import RedisObject, namespace_lock
+if TYPE_CHECKING:
+    from ._cog import DocItem
+
+
+class DocRedisCache(RedisObject):
+    """Interface for redis functionality needed by the Doc cog."""
+
+    @namespace_lock
+    async def set(self, item: DocItem, value: str) -> None:
+        """
+        Set markdown `value` for `key`.
+
+        Keys expire after a week to keep data up to date.
+        """
+        expiry_timestamp = datetime.datetime.now().timestamp() + 7 * 24 * 60 * 60
+        with await self._get_pool_connection() as connection:
+            await connection.hset(
+                f"{self.namespace}:{item.package}",
+                self.get_item_key(item),
+                pickle.dumps((value, expiry_timestamp))
+            )
+
+    @namespace_lock
+    async def set_if_exists(self, item: DocItem, value: str) -> None:
+        """
+        Set markdown `value` for `key` if `key` exists.
+
+        Keys expire after a week to keep data up to date.
+        """
+        expiry_timestamp = datetime.datetime.now().timestamp() + 7 * 24 * 60 * 60
+
+        with await self._get_pool_connection() as connection:
+            if await connection.hexists(f"{self.namespace}:{item.package}", self.get_item_key(item)):
+                await connection.hset(
+                    f"{self.namespace}:{item.package}",
+                    self.get_item_key(item),
+                    pickle.dumps((value, expiry_timestamp))
+                )
```

Comment:
I couldn't really think of anything that wouldn't unnecessarily complicate this because we also need to hold the connection  within the context manager and the conditional is in the middle of it